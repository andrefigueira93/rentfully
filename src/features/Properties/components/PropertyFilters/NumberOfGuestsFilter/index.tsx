import { Button } from "@/components/ui/button";
import { numberOfGuestsFilterAtom } from "@/store/booking-filters-store";
import { maxGuestsAtom } from "@/store/properties-store";
import { MinusIcon, PlusIcon } from "@radix-ui/react-icons";
import { useAtom, useAtomValue } from "jotai";
import { toast } from "sonner";
import FilterWrapper from "../FilterWrapper";

export default function NumberOfGuestsFilter() {
  const [numberOfGuests, setNumberOfGuests] = useAtom(numberOfGuestsFilterAtom);
  const maxGuests = useAtomValue(maxGuestsAtom);

  function SubGuest() {
    setNumberOfGuests((prev) => {
      if (prev > 1) {
        return prev - 1;
      }

      toast("Minimium of one guest", {
        cancel: {
          label: "Ok",
        },
      });

      return prev;
    });
  }

  function AddGuest() {
    setNumberOfGuests((prev) => {
      if (prev < maxGuests) {
        return prev + 1;
      }
      toast("We do not have enough rooms for this number of guests", {
        cancel: {
          label: "Ok",
        },
      });
      return prev;
    });
  }

  return (
    <FilterWrapper label="Number of guests">
      <Button
        aria-labelledby="Subtract guest"
        data-testid="subtract-guest"
        onClick={SubGuest}
        variant="outline"
      >
        <MinusIcon />
      </Button>
      <span className="mx-2">{numberOfGuests}</span>
      <Button data-testid="add-guest" onClick={AddGuest} variant="outline">
        <PlusIcon />
      </Button>
    </FilterWrapper>
  );
}
