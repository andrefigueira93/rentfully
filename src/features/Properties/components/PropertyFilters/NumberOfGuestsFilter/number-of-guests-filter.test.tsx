/* eslint-disable @typescript-eslint/no-explicit-any */
import Providers from "@/providers";
import { fireEvent, render, screen } from "@testing-library/react";
import { useAtom } from "jotai";
import { MemoryRouter } from "react-router";
import NumberOfGuestsFilter from "./index";

vi.mock("jotai", async () => {
  const originalModule = await import("jotai");

  return {
    ...originalModule,
    useAtom: vi.fn(),
  };
});

const mockSetValue = vi.fn() as never;
const selectedGuests = 5;

const mockSetNumberOfGuests = vi.fn() as never;

describe("NumberOfGuestsFilter Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useAtom).mockReturnValue([selectedGuests, mockSetValue]);
  });

  test("renders correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <Providers>
          <NumberOfGuestsFilter />
        </Providers>
      </MemoryRouter>
    );

    expect(screen.getByText("Number of guests")).toBeInTheDocument();
    expect(screen.getByText("5")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("subtracts a guest", () => {
    vi.mocked(useAtom).mockReturnValue([5, mockSetNumberOfGuests]);

    const { container } = render(
      <MemoryRouter>
        <Providers>
          <NumberOfGuestsFilter />
        </Providers>
      </MemoryRouter>
    );

    const subtractButton = screen.getByTestId("subtract-guest");
    fireEvent.click(subtractButton);

    expect(mockSetNumberOfGuests).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test("adds a guest", () => {
    vi.mocked(useAtom).mockReturnValue([10, mockSetNumberOfGuests]);

    const { container } = render(
      <MemoryRouter>
        <Providers>
          <NumberOfGuestsFilter />
        </Providers>
      </MemoryRouter>
    );

    const addButton = screen.getByTestId("add-guest");
    fireEvent.click(addButton);

    expect(mockSetNumberOfGuests).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
