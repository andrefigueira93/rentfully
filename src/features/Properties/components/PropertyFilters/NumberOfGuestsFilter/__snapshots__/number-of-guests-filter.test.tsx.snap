// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumberOfGuestsFilter Component > adds a guest 1`] = `
<div>
  <div
    class="flex flex-col space-y-2"
  >
    <div
      class="w-full flex items-center justify-evenly"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-full"
      >
        Number of guests
      </label>
    </div>
    <div
      class="w-full flex items-center justify-evenly"
    >
      <button
        aria-labelledby="Subtract guest"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="subtract-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.25 7.5C2.25 7.22386 2.47386 7 2.75 7H12.25C12.5261 7 12.75 7.22386 12.75 7.5C12.75 7.77614 12.5261 8 12.25 8H2.75C2.47386 8 2.25 7.77614 2.25 7.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <span
        class="mx-2"
      >
        10
      </span>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="add-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8 2.75C8 2.47386 7.77614 2.25 7.5 2.25C7.22386 2.25 7 2.47386 7 2.75V7H2.75C2.47386 7 2.25 7.22386 2.25 7.5C2.25 7.77614 2.47386 8 2.75 8H7V12.25C7 12.5261 7.22386 12.75 7.5 12.75C7.77614 12.75 8 12.5261 8 12.25V8H12.25C12.5261 8 12.75 7.77614 12.75 7.5C12.75 7.22386 12.5261 7 12.25 7H8V2.75Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`NumberOfGuestsFilter Component > renders correctly 1`] = `
<div>
  <div
    class="flex flex-col space-y-2"
  >
    <div
      class="w-full flex items-center justify-evenly"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-full"
      >
        Number of guests
      </label>
    </div>
    <div
      class="w-full flex items-center justify-evenly"
    >
      <button
        aria-labelledby="Subtract guest"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="subtract-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.25 7.5C2.25 7.22386 2.47386 7 2.75 7H12.25C12.5261 7 12.75 7.22386 12.75 7.5C12.75 7.77614 12.5261 8 12.25 8H2.75C2.47386 8 2.25 7.77614 2.25 7.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <span
        class="mx-2"
      >
        5
      </span>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="add-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8 2.75C8 2.47386 7.77614 2.25 7.5 2.25C7.22386 2.25 7 2.47386 7 2.75V7H2.75C2.47386 7 2.25 7.22386 2.25 7.5C2.25 7.77614 2.47386 8 2.75 8H7V12.25C7 12.5261 7.22386 12.75 7.5 12.75C7.77614 12.75 8 12.5261 8 12.25V8H12.25C12.5261 8 12.75 7.77614 12.75 7.5C12.75 7.22386 12.5261 7 12.25 7H8V2.75Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`NumberOfGuestsFilter Component > subtracts a guest 1`] = `
<div>
  <div
    class="flex flex-col space-y-2"
  >
    <div
      class="w-full flex items-center justify-evenly"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-full"
      >
        Number of guests
      </label>
    </div>
    <div
      class="w-full flex items-center justify-evenly"
    >
      <button
        aria-labelledby="Subtract guest"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="subtract-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M2.25 7.5C2.25 7.22386 2.47386 7 2.75 7H12.25C12.5261 7 12.75 7.22386 12.75 7.5C12.75 7.77614 12.5261 8 12.25 8H2.75C2.47386 8 2.25 7.77614 2.25 7.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <span
        class="mx-2"
      >
        5
      </span>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2"
        data-testid="add-guest"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8 2.75C8 2.47386 7.77614 2.25 7.5 2.25C7.22386 2.25 7 2.47386 7 2.75V7H2.75C2.47386 7 2.25 7.22386 2.25 7.5C2.25 7.77614 2.47386 8 2.75 8H7V12.25C7 12.5261 7.22386 12.75 7.5 12.75C7.77614 12.75 8 12.5261 8 12.25V8H12.25C12.5261 8 12.75 7.77614 12.75 7.5C12.75 7.22386 12.5261 7 12.25 7H8V2.75Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
